share/emock.sh

```diff
@@ -0,0 +1,434 @@
+
+            # Return
+            echo -n '${return_code}' > '${statedir}'/${called}/return_code
+            return '${return_code}'
+        }'
+    else
+        body='
+        {
+            '${base_body}'
+            ( '${body}' )
+
+            # Return
+            local return_code=$?
+            echo -n ${return_code} > '${statedir}'/${called}/return_code
+            return ${return_code}
+        }'
+    fi
+
+    # Create _real function wrapper to call the real, unmodified function
+    edebug "Creating ${name}_real function"
+    if command ${name} &>/dev/null; then
+        eval "${name}_real () { command ${name} \"\${@}\"; }"
+    else
+        local real
+        real="$(declare -f ${name})"
+        eval "${name}_real${real#${name}}"
+    fi
+
+    eval "declare -f ${name}_real"
+
+    # Create a function wrapper to call our mock function instead of the real function.
+    edebug "Creating ${name} mock function with $(lval body)"
+    eval "${name} () ${body}"
+    eval "declare -f ${name}"
+}
+
+opt_usage unmock <<'END'
+eunmock is used in tandem with emock to remove a mock that has previosly been created. This essentially removes the
+wrapper functions we create and also cleans up the on-disk statedir for the mock.
+END
+eunmock()
+{
+    $(opt_parse \
+        ":statedir=.emock                   | This directory is used to track state about mocked binaries. This will
+                                              hold metadata information such as the number of times the mock was called
+                                              as well as the exit code, stdout, and stderr for each invocation."       \
+        "name                               | Name of the binary to mock (e.g. dmidecode or /usr/sbin/dmidecode)."     \
+    )
+
+    edebug "Removing ${name} mock function"
+    unset -f "${name}_real"
+    unset -f "${name}"
+
+    edebug "Removing ${name} mock state"
+    rm -rf "${statedir}/$(basename "${name}")"
+}
+
+opt_usage emock_called <<'END'
+emock_called make it easier to check how many times a mock has been called. This is tracked on-disk in the statedir in
+the file named "called". While it's easy to manually retrieve from this file, this function should always be used to
+provide a clean abstraction.
+
+Just like a typical array in any language, the size, or count of the number of times that the
+mock has been called is 1-based but the actual index values we use to store the state files for each invocation is
+zero-based (again, just like an array).
+
+So if this has never been called, then emock_called will return 0, and there will be no on-disk state directory. The
+first time you call it, emock_called will return 1, and there will be a ${statedir}/0 directory storing the state files
+for that invocation.
+END
+emock_called()
+{
+    $(opt_parse \
+        ":statedir=.emock                   | This directory is used to track state about mocked binaries. This will
```

Comment:
Something to consider: have the statedir be something like ".emock-${pid}" so you can parallelize stuff using emock

Reply:
Ooo, I like this idea!! great suggestion @ericpe!

Reply:
OH man, that is a great idea. I've implemented this!! Thanks for the suggestion @ericpe